File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.java
import java.io.FileInputStream;
import java.io.IOException;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.Deque;
import java.util.Scanner;

public class Main {
	static ArrayList<Integer>[] graph;
	static int n;
	static boolean[] visited;
	public static void main(String[] args) throws IOException {
		//System.setIn(new FileInputStream("data/input.txt"));
		Scanner sc = new Scanner(System.in);
		n = sc.nextInt();
		int m = sc.nextInt();
		int v = sc.nextInt();
		graph = new ArrayList[n+1];
		for (int i = 1; i <= n; i++) {
			graph[i] = new ArrayList<>();
		}
		for (int i = 0; i < m; i++) {
			int a = sc.nextInt();
			int b = sc.nextInt();
			graph[a].add(b);
			graph[b].add(a);
		}
		for (int i = 1; i <= n; i++) {
			Collections.sort(graph[i]);
		}

		visited = new boolean[n+1];
		visited[v] = true;
		System.out.print(v + " ");
		dfs(v);
		System.out.println();
		visited = new boolean[n+1];
		visited[v] = true;
		System.out.print(v + " ");
		
		bfs(v);
	}
	private static void dfs(int v) {
		for (int next : graph[v]) {
			if (visited[next]) continue;
			visited[next] = true;
			System.out.print(next + " ");
			dfs(next);
		}
	}
	private static void bfs(int v) {
		Deque<Integer> queue = new ArrayDeque<>();
		queue.offer(v);
		while (!queue.isEmpty()) {
			int cur = queue.poll();
			for (int next : graph[cur]) {
				if (visited[next]) continue;
				visited[next] = true;
				System.out.print(next + " ");
				queue.offer(next);
			}
		}
		
	}
}

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
'''
4 5 1
1 2
1 3
1 4
2 4
3 4
'''

import sys
from collections import deque

'''
주의사항
두 정점 사이에 간선이 여러개일 수 있음
'''

input = sys.stdin.readline

n, m, v = map(int, input().split())
graph = [[] for _ in range(n+1)]
for i in range(m):
    a, b = map(int, input().rstrip().split())
    graph[a].append(b)
    graph[b].append(a)

for i in range(len(graph)):
    graph[i].sort()


def dfs(v, graph, visited):
    visited[v] = True
    print(v, end=' ')
    for i in graph[v]:
        if not visited[i]:
            dfs(i, graph, visited)


def bfs(v, graph, visited):
    queue = deque([v])
    visited[v] = True
    while queue:
        v = queue.popleft()
        print(v, end=" ")
        for i in graph[v]:
            if not visited[i]:
                queue.append(i)
                visited[i] = True


visited = [False] * (n+1)
dfs(v, graph, visited)
print("\n",end='')
visited = [False] * (n+1)
bfs(v, graph, visited)



